File: hamili-app/lib/features/budgets/domain/budget.dart
class AppBudget {
  final int id;
  final int categoryId;
  final int month;
  final int year;
  final double limitAmount;
  final double spentAmount;
  final double remainingAmount;
  final double percentageUsed;

  const AppBudget({
    required this.id,
    required this.categoryId,
    required this.month,
    required this.year,
    required this.limitAmount,
    required this.spentAmount,
    required this.remainingAmount,
    required this.percentageUsed,
  });

  factory AppBudget.fromJson(Map<String, dynamic> json) => AppBudget(
        id: json['id'] as int,
        categoryId: json['category_id'] as int,
        month: json['month'] as int,
        year: json['year'] as int,
        limitAmount: (json['limit_amount'] as num).toDouble(),
        spentAmount: (json['spent_amount'] as num).toDouble(),
        remainingAmount: (json['remaining_amount'] as num).toDouble(),
        percentageUsed: (json['percentage_used'] as num).toDouble(),
      );
}

File: hamili-backend/app/schemas/budget.py
from pydantic import BaseModel, Field


class BudgetCreate(BaseModel):
    """POST /budgets is an upsert — creating a budget for a category/month/
    year that already has one updates the existing row instead of erroring,
    since 'set my Food budget to ₱3000' should just work whether or not
    you'd set it before."""

    category_id: int
    month: int = Field(ge=1, le=12)
    year: int = Field(ge=2020, le=2100)
    limit_amount: float = Field(gt=0, le=1_000_000_000_000)


class BudgetUpdate(BaseModel):
    limit_amount: float = Field(gt=0, le=1_000_000_000_000)


class BudgetOut(BaseModel):
    id: int
    category_id: int
    month: int
    year: int
    limit_amount: float
    spent_amount: float
    remaining_amount: float
    percentage_used: float
